oedisi/tools/metrics.py

```diff
@@ -0,0 +1,162 @@
+
+    The measurements are assumed to be in the form of .feather files.
+
+    \b
+    True voltages:
+    - voltage_real.feather
+    - voltage_imag.feather
+
+    \b
+    Estimated voltages
+    - voltage_mag.feather
+    - voltage_angle.feather
+
+    \b
+    Metrics:
+    - MARE: Mean absolute relative error.
+    - RMSRE: Root mean squared relative error.
+    - MAAE: Mean absolute angle error.
+
+    \f
+    Parameters
+    ----------
+
+    path : Path
+        Path to the folder containing the measurement files.
+
+    metric : str
+        Metric to be used for evaluation. The options are:
+    """
+    path = Path(path)
+    if not _has_dependencies:
+        raise ImportError("numpy and pandas are required to do this.")
+    # measurement files have columns with buses along with a time column
+    true_voltages_real = pd.read_feather(path / "voltage_real.feather")
+    true_voltages_imag = pd.read_feather(path / "voltage_imag.feather")
+
+    estimated_magnitude = pd.read_feather(path / "voltage_mag.feather")
+    estimated_angle = pd.read_feather(path / "voltage_angle.feather")
+
+    time = true_voltages_real["time"]
+    estimated_time = estimated_magnitude["time"]
+    common_time = set(time).intersection(estimated_time)
+
+    true_voltages_real = true_voltages_real[
+        true_voltages_real["time"].isin(common_time)
+    ]
+    true_voltages_imag = true_voltages_imag[
+        true_voltages_imag["time"].isin(common_time)
+    ]
+    estimated_magnitude = estimated_magnitude[
+        estimated_magnitude["time"].isin(common_time)
+    ]
+    estimated_angle = estimated_angle[estimated_angle["time"].isin(common_time)]
+
+    estimated_magnitude = estimated_magnitude.groupby("time").last().reset_index()
+    estimated_angle = estimated_angle.groupby("time").last().reset_index()
+
+    assert list(true_voltages_real["time"]) == list(
+        estimated_magnitude["time"]
+    ), f"""Time does not match between true voltages and estimated magnitudes:
+    
+    {list(true_voltages_real["time"])} vs {list(estimated_magnitude["time"])}"""
+
+    # Strip time column from voltages
+    true_voltages_real = true_voltages_real.drop(columns=["time"]).reset_index(
+        drop=True
+    )
+    true_voltages_imag = true_voltages_imag.drop(columns=["time"]).reset_index(
+        drop=True
+    )
+    estimated_magnitude = estimated_magnitude.drop(columns=["time"]).reset_index(
+        drop=True
+    )
+    estimated_angle = estimated_angle.drop(columns=["time"]).reset_index(drop=True)
+    if angle_unit == "degrees":  # convert to radians
+        estimated_angle = estimated_angle * np.pi / 180
+
+    # Convert true voltages to magnitude and angles
+    true_magnitudes = np.hypot(true_voltages_real, true_voltages_imag)
+    true_angles = np.arctan2(true_voltages_imag, true_voltages_real)
+
+    nonzero_parts = true_magnitudes != 0
+    assert (
+        true_magnitudes.columns == estimated_magnitude.columns
```

Comment:
question: should this use an isclose() type of computation? Or is it ok to evaluate the values exactly?